### PR TITLE
*: serve '/metrics' in insecure port

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/8226.

### CODE REVIEW

File: embed/config.go [hunk 1]
```diff
@@ -118,6 +118,7 @@ type Config struct {
 	Metrics               string `json:"metrics"`
 	ListenMetricsUrls     []url.URL
 	ListenMetricsUrlsJSON string `json:"listen-metrics-urls"`
+	ListenMetricsInsecure bool   `json:"listen-metrics-insecure"`
```
Comment (heyitsanthony) [inline embed/config.go]: why is this needed? this should probably be like the listen flags so it accepts "https://whatever,http://whatever" to use insecure connections with http

**Comment by gyuho** [comment]
@heyitsanthony PTAL.

Will add health false with alarm list field in the next PR.

**DO NOT MERGE UNTIL https://github.com/coreos/etcd/pull/8277**

File: etcdserver/api/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	mux.Handle(PathMetrics, prometheus.Handler())
+}
+
+// RegisterHealth registers health handler on '/health'.
+func RegisterHealth(mux *http.ServeMux, srv *etcdserver.EtcdServer) {
+	mux.Handle(PathHealth, HealthHandler(srv))
+}
+
+// Handler serves health and metrics information.
+func Handler(srv *etcdserver.EtcdServer) http.HandlerFunc {
```
Comment (heyitsanthony) [inline etcdserver/api/metrics/metrics.go]: `NewMetricsHandler` to be consistent with `NewClientHandler` in v2http?

File: etcdserver/api/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+			http.Error(w, `{"health": "false"}`, http.StatusServiceUnavailable)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(`{"health": "true"}`))
+	}
+}
+
+// RegisterPrometheus registers prometheus handler on '/metrics'.
+func RegisterPrometheus(mux *http.ServeMux) {
```
Comment (heyitsanthony) [inline etcdserver/api/metrics/metrics.go]: should these be called HandlePrometheus and HandleHealth since there's already Handle everywhere?

File: etcdserver/api/metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+	"github.com/coreos/etcd/etcdserver"
+	"github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/raft"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	PathMetrics = "/metrics"
```
Comment (heyitsanthony) [inline etcdserver/api/metrics/metrics.go]: keep unexported until the code needs to override them?

File: etcdserver/api/metrics/metrics.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
+// Copyright 2017 The etcd Authors
```
Comment (heyitsanthony) [inline etcdserver/api/metrics/metrics.go]: can this stay in etcdhttp but just export the handlers? having both health and metrics in a package called metrics is a little confusing (e.g., does Handler give me a metrics handler or health+metrics handlers)

File: e2e/cluster_proxy_test.go [hunk 6]
```diff
@@ -55,6 +55,9 @@ func (p *proxyEtcdProcess) Config() *etcdServerProcessConfig { return p.etcdProc
 func (p *proxyEtcdProcess) EndpointsV2() []string { return p.proxyV2.endpoints() }
 func (p *proxyEtcdProcess) EndpointsV3() []string { return p.proxyV3.endpoints() }
 
+// TODO: proxy doesn't provide health information
+func (p *proxyEtcdProcess) EndpointsMetrics() []string { return []string{p.proxyV3.murl} }
```
Comment (heyitsanthony) [inline e2e/cluster_proxy_test.go]: can this panic instead? the metrics tests are !cluster_proxy so it shouldn't hit this path

File: e2e/metrics_test.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	"github.com/coreos/etcd/pkg/testutil"
+)
+
+func TestV3MetricsSecure(t *testing.T)   { testV3Metrics(t, false) }
+func TestV3MetricsInsecure(t *testing.T) { testV3Metrics(t, true) }
+func testV3Metrics(t *testing.T, insecureMetricsURL bool) {
+	defer testutil.AfterTest(t)
+
+	cfg := configTLS
+	cfg.metricsURL = true
```
Comment (heyitsanthony) [inline e2e/metrics_test.go]: possible to simplify the config by replacing the `metricsURL` and `insecureMetricsURL` with `metricsURLScheme string` with empty string meaning no metrics url:
```go
func TestV3MetricsSecure(t *testing.T)   { testV3Metrics(t, "https") }
func testV3Metrics(t *testing.T, scheme string) {
	defer testutil.AfterTest(t)

	cfg := configTLS
	cfg.metricsURLScheme = scheme
        ...
}
```

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Thanks.

File: etcdserver/api/etcdhttp/metrics.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	pathMetrics = "/metrics"
+	pathHealth  = "/health"
+)
+
+// Health returns true if etcdserver is available.
+func Health(srv *etcdserver.EtcdServer) bool {
```
Comment (heyitsanthony) [inline etcdserver/api/etcdhttp/metrics.go]: unexport?

File: etcdserver/api/etcdhttp/metrics.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+		return false
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	_, err := srv.Do(ctx, etcdserverpb.Request{Method: "QGET"})
+	cancel()
+	return err == nil
+}
+
+// HealthHandler handles '/health' requests.
+func HealthHandler(srv *etcdserver.EtcdServer) http.HandlerFunc {
```
Comment (heyitsanthony) [inline etcdserver/api/etcdhttp/metrics.go]: `NewHealthHandler`

File: etcdserver/api/etcdhttp/metrics.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+// HandleHealth registers health handler on '/health'.
+func HandleHealth(mux *http.ServeMux, srv *etcdserver.EtcdServer) {
+	mux.Handle(pathHealth, HealthHandler(srv))
+}
+
+// NewMetricsHandler serves health and metrics information.
+func NewMetricsHandler(srv *etcdserver.EtcdServer) http.HandlerFunc {
+	h := prometheus.Handler()
+	return func(w http.ResponseWriter, r *http.Request) {
```
Comment (heyitsanthony) [inline etcdserver/api/etcdhttp/metrics.go]: reuse the health handler instead of reimplementing?
```go
hh := NewHealthHander(srv)
return func(...) {
    ...
    hh(w, r)
}
```

File: etcdserver/api/etcdhttp/metrics.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+		hj, ok := w.(http.Hijacker)
+		if !ok { // webserver doesn't support hijacking
+			h.ServeHTTP(w, r)
+			return
+		}
+		conn, buf, err := hj.Hijack()
```
Comment (heyitsanthony) [inline etcdserver/api/etcdhttp/metrics.go]: probably needs a comment on why this needs hijacking; it's not clear to me why ordinary muxing won't work

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Thanks!

File: etcdserver/api/etcdhttp/metrics.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+// HandleHealth registers health handler on '/health'.
+func HandleHealth(mux *http.ServeMux, srv *etcdserver.EtcdServer) {
+	mux.Handle(pathHealth, newHealthHandler(srv, "json"))
+}
+
+// NewMetricsHandler serves health and metrics information.
+func NewMetricsHandler(srv *etcdserver.EtcdServer) http.HandlerFunc {
+	h := prometheus.Handler()
+	hh := newHealthHandler(srv, "prometheus") // to be consistent with Prometheus handler
```
Comment (heyitsanthony) [inline etcdserver/api/etcdhttp/metrics.go]: Shouldn't the metrics port serve "/metrics" and "/health" with separate handlers so it matches the routes normally exposed by etcd? Like this:
```go
func HandleMetricsHealth(mux *http.ServeMux, srv *etcdserver.EtcdServer) {
    mux.Handler(pathMetrics, NewMetricsHandler(srv))
    mux.Handler(pathHealth, NewHealthHandler(srv))
}
```

**Comment by gyuho** [comment]
@heyitsanthony Just used `HandleMetricsHealth` making `NewMetricsHandler` unexported.

Thanks.

File: etcdserver/api/etcdhttp/metrics.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+func HandleHealth(mux *http.ServeMux, srv *etcdserver.EtcdServer) {
+	mux.Handle(pathHealth, newHealthHandler(srv, "json"))
+}
+
+// newMetricsHandler serves health and metrics information.
+func newMetricsHandler(srv *etcdserver.EtcdServer) http.HandlerFunc {
+	h := prometheus.Handler()
+	hh := newHealthHandler(srv, "prometheus") // to be consistent with Prometheus handler
+	return func(w http.ResponseWriter, r *http.Request) {
+		hh(w, r)
```
Comment (heyitsanthony) [inline etcdserver/api/etcdhttp/metrics.go]: why is the handler still writing out the health if it can be accessed through `/health`? the metrics should probably match an ordinary client endpoint if possible...

File: e2e/metrics_test.go [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+	if err != nil {
+		t.Fatalf("could not start etcd process cluster (%v)", err)
+	}
+	defer func() {
+		if cerr := epc.Close(); err != nil {
+			t.Fatalf("error closing etcd processes (%v)", cerr)
+		}
+	}()
+
+	if err = cURLGet(epc, cURLReq{endpoint: "/metrics", expected: `health true`, metricsURL: true}); err != nil {
```
Comment (heyitsanthony) [inline e2e/metrics_test.go]: test /health as well as /metrics?

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Thanks.
